// @purduesigbots/pros-cli-middleware/dist/input/interactive/components/Container.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Component_1 = require("./Component");
var Container = /** @class */ (function (_super) {
    __extends(Container, _super);
    function Container(_a) {
        var title = _a.title, description = _a.description, uuid = _a.uuid, collapsed = _a.collapsed, elements = _a.elements, _update = _a._update, __componentFactory = _a.__componentFactory;
        var _this = _super.call(this) || this;
        _this.title = title;
        _this.description = description;
        _this.uuid = uuid;
        _this.collapsed = collapsed;
        _this.elements = elements.map(function (e) { return __componentFactory.createInstance(e); });
        _this.update_callback = _update;
        return _this;
    }
    Container.prototype.updateCollapsed = function (collapsed) {
        this.update_callback({
            event: 'update',
            uuid: this.uuid,
            args: [collapsed]
        });
        this.collapsed = collapsed;
    };
    return Container;
}(Component_1.Component));
exports.Container = Container;
